File: csharp-playwright-nunit-reqnroll/SauceDemo.UiTests/Users/UserLease.cs
namespace SauceDemo.UiTests.Users;

/// <summary>Holds a leased <see cref="UserAccount"/>; disposing returns it to its pool exactly once.</summary>
public sealed class UserLease : IDisposable
{
    private readonly Action<UserAccount>? _release;
    private bool _disposed;

    internal UserLease(UserAccount account, Action<UserAccount> release)
    {
        Account = account;
        _release = release;
    }

    /// <summary>
    /// Wraps an account that was not leased from a pool (e.g. a tag-based `@user:` override that
    /// reads directly from <see cref="UserCatalog"/>). Disposing it is a no-op, since there is no
    /// pool to return the account to.
    /// </summary>
    public UserLease(UserAccount account)
    {
        Account = account;
        _release = null;
    }

    public UserAccount Account { get; }

    public void Dispose()
    {
        if (_disposed)
        {
            return;
        }

        _disposed = true;
        _release?.Invoke(Account);
    }
}

File: java-selenium-junit5-cucumber/src/test/java/com/saucedemo/uitests/users/UserLease.java
package com.saucedemo.uitests.users;

import java.util.function.Consumer;

/** Holds a leased {@link UserAccount}; closing returns it to its pool exactly once. */
public final class UserLease implements AutoCloseable {

    private final UserAccount account;
    private final Consumer<UserAccount> release;
    private boolean closed;

    UserLease(UserAccount account, Consumer<UserAccount> release) {
        this.account = account;
        this.release = release;
    }

    /**
     * Wraps an account that was never leased from a pool - a {@code @user:<username>} tag override that
     * reads straight from {@link UserCatalog}. Closing it is a no-op, because there is no pool to return
     * it to.
     */
    public UserLease(UserAccount account) {
        this(account, null);
    }

    public UserAccount account() {
        return account;
    }

    @Override
    public void close() {
        if (closed) {
            return;
        }
        closed = true;

        if (release != null) {
            release.accept(account);
        }
    }
}
